test/validation/api/ipsec/ipsec.c

```diff
@@ -974,6 +982,23 @@ void ipsec_check_in_one(const ipsec_test_part *part, odp_ipsec_sa_t sa)
 	verify_in_one(part, sa, pkto);
 }
 
+void ipsec_out_verify_send(const ipsec_test_part *part,
+			   odp_ipsec_sa_t sa,
+			   int num_part)
+{
+	ipsec_test_packet pkt_in[num_part];
+	ipsec_test_part part_in[num_part];
+	odp_packet_t pkto[num_part];
+	int i;
+
+	for (i = 0; i < num_part; i++) {
+		ipsec_check_out(&part[i], sa, &pkto[i]);
+		ipsec_prep_part_from_pkt(&part_in[i], &pkto[i], &pkt_in[i]);
+	}
+
+	send_pkts(part_in, i);
```

Comment:
I am not sure if this is a good way. Maybe the primitives provided by ipsec.c should be only about prosessing one packet or a set of related packets and checking the result. And all other logic could be in the test cases. Maybe this function (and ipsec_in_recv_verify too) should not be in ipsec.c by near the test cases and only do the outbound processing.

I was thinking that ipsec_check_in_one() (which could perhaps be called ipsec_process_inbound()) could take single ipsec_test_part which could contain an array of input packets and an expected result packet. Test case code could then do ipsec_check_out() for multiple packets (as it does not) and then one call to ipsec_check_in_one() to process the ESP packets and check the result. In case of incomplete reassembly, maybe the test case code could do the checking of the received fragments (?).

Reply:
> I am not sure if this is a good way. Maybe the primitives provided by ipsec.c should be only about prosessing one packet or a set of related packets and checking the result. And all other logic could be in the test cases. Maybe this function (and ipsec_in_recv_verify too) should not be in ipsec.c by near the test cases and only do the outbound processing.
> 

Well, how do you judge send_pkts() should be part of ipsec_test_out.c or in ipsec_test_in.c without looking at the callers? Or why do you say that function should be called only from out & not from in (and vice-versa?). Also, if your concern is about adding everything into ipsec.c, I can move it into separate header but I really don't see any point in doing so. The only reason why I introduced these small functions is to minimize code duplication. If you can confirm that we are okay with code duplication, I can take that direction.

> I was thinking that ipsec_check_in_one() (which could perhaps be called ipsec_process_inbound()) could take single ipsec_test_part which could contain an array of input packets and an expected result packet. 

I thought we decided to not have one test part have multiple input packets? IMO, this kind of abstraction is what made the UT very complicated at the first place and you had cleaned some part of it recently. I had confirmed all these aspects then itself.

> Test case code could then do ipsec_check_out() for multiple packets (as it does not) and then one call to ipsec_check_in_one() to process the ESP packets and check the result. In case of incomplete reassembly, maybe the test case code could do the checking of the received fragments (?).

Can you highlight the issues with current approach? If you have something specific in mind, can you share code snippets?

Reply:
@JannePeltonen To summarize the approach that I took,

1. Identified pieces that are common for reassembly and non-reassembly paths
2. Made that common pieces into a separate functions (same file, static)
3. Made new functions for reassembly use case calling the ones identified in Step 2 without having any modification in non-reassembly paths
4. Introduce reassembly tests calling the above.